src/main/java/com/google/devtools/build/lib/remote/GrpcRemoteOutputService.java

```diff
@@ -0,0 +1,527 @@
+      // TODO: Do we need to partition the input, just like in
+      // createSymlinkTree(), or is input already guaranteed to be
+      // bounded in size?
+      BatchStatRequest.Builder builder = BatchStatRequest.newBuilder();
+      builder.setBuildId(stateAcrossBuilds.currentBuildId.toString());
+      builder.setIncludeFileDigest(includeDigest);
+      for (PathFragment path : paths) {
+        builder.addPaths(fixupExecRootPath(path));
+      }
+
+      BatchStatResponse responses;
+      try {
+        responses = blockingStub.batchStat(builder.build());
+      } catch (StatusRuntimeException e) {
+        throw new IOException(e);
+      }
+      return Lists.newArrayList(
+          Iterables.transform(
+              responses.getResponsesList(),
+              (response) -> {
+                if (!response.hasFileStatus()) {
+                  // File not found.
+                  return null;
+                }
+                FileStatus fileStatus = response.getFileStatus();
+                if (fileStatus.hasFile()) {
+                  Digest digest = fileStatus.getFile().getDigest();
+                  return new RegularFileStatus(digest.getSizeBytes(), DigestUtil.toBinaryDigest(digest));
+                }
+                if (fileStatus.hasDirectory()) {
+                  throw new RuntimeException("DIRECTORY");
+                }
+                if (fileStatus.hasSymlink()) {
+                  return new SymlinkFileStatus();
+                }
+                if (fileStatus.hasExternal()) {
+                  throw new RuntimeException("EXTERNAL");
+                }
+                throw new RuntimeException("SOMETHING ELSE");
+              }));
+    }
+  }
+
+  @Override
+  public BatchStat getBatchStatter() {
+    return new GrpcBatchStat();
+  }
+
+  @Override
+  public boolean canCreateSymlinkTree() {
+    return true;
+  }
+
+  @Override
+  public void createSymlinkTree(Map<PathFragment, PathFragment> symlinks, PathFragment symlinkTreeRoot)
+      throws ExecException, InterruptedException {
+    // The provided set of symbolic links may be too large to provide to
+    // the remote output service at once. Partition the symbolic links
+    // in groups of 1000, so that BatchCreateRequest messages remain
+    // small enough.
+    UnmodifiableIterator<List<Map.Entry<PathFragment, PathFragment>>> symlinkBatchIterator =
+        partition(
+            filter(symlinks.entrySet().iterator(), (symlink) -> symlink.getValue() != null),
+            1000);
+    boolean cleanPathPrefix = true;
+    while (symlinkBatchIterator.hasNext()) {
+      List<Map.Entry<PathFragment, PathFragment>> symlinksBatch = symlinkBatchIterator.next();
+      BatchCreateRequest.Builder builder = BatchCreateRequest.newBuilder();
+      builder.setBuildId(stateAcrossBuilds.currentBuildId.toString());
+      builder.setPathPrefix(symlinkTreeRoot.toString());
+      builder.setCleanPathPrefix(cleanPathPrefix);
```

Comment:
What does `cleanPathPrefix` stand for? In here it is first batch, really.

Reply:
It causes the existing contents of the symlink tree directory to be removed. I have just added a comment for this.

Reply:
Did you consider using a streaming API? In that case, cleanup on could would be implicit. Not sure what's easier, also not sure how large do those requests get.